### PR TITLE
Adding shared template option

### PR DESCRIPTION
**Pull request by igor-ribeiiro**
Thank you for your pull request. Please provide the following information.

---

1. **Do these changes impact *User Experience*?** (e.g., how the user interacts with Yo Office and/or the files and folders the user sees in the project that Yo Office creates)
    > 
    > * [X]  Yes
    > * [ ]  No

There will be a new option for the shared runtime template.

2. **Do these changes impact *documentation*?** (e.g., a tutorial on https://docs.microsoft.com/en-us/office/dev/add-ins/overview/office-add-ins)
    > 
    > * [X]  Yes
    > * [ ]  No

The tutorial on how to create CFs need to be updated. It also needs to include shared vs legacy runtimes differences.

3. **Validation/testing performed**:

Tested creating a new yo office app and choosing the option shared template. The template is indeed the shared one.

4. **Platforms tested**:

    > * [X] Windows
    > * [ ] Mac

### CODE REVIEW

File: src/app/config/projectProperties.json [hunk 1]
```diff
@@ -129,6 +129,25 @@
                 "Excel"
             ]
         },
+        "excel-functions-shared": {
+            "displayname": "Excel Shared Runtime Custom Functions Add-in project",
```
[the remark this answers is not in the document]
Comment (igor-ribeiiro) [reply src/app/config/projectProperties.json:133]: I don't think there's a name yet. I will keep this PR on hold until we get the name confirmation and then change when we have it

[the remark this answers is not in the document]
Comment (Rick-Kirkham) [reply src/app/config/projectProperties.json:133]: @alison-mk can help choose a name. If it were me, I'd like to avoid so many stacked nouns. I'd prefer breaking it up with a prepositional phrase. "Excel Custom Functions with a Shared Runtime project".

**Review by Rick-Kirkham — APPROVED** [review]

File: src/app/config/projectProperties.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (akrantz) [reply src/app/config/projectProperties.json:133]: "Excel Custom Functions using Shared Runtime"  ?

[the remark this answers is not in the document]
Comment (igor-ribeiiro) [reply src/app/config/projectProperties.json:133]: We can also change the normal Excel CFs name. Please let me know if we need to do that.

[the remark this answers is not in the document]
Comment (alison-mk) [reply src/app/config/projectProperties.json:133]: I like Ricky and Adam's suggestions. Could we also place Shared Runtime before JS-only in the list, to subtly indicate that it's the preferred method? @igor-ribeiiro Yes, I would also like to change the existing CF template name. I recommend: 

- Excel Custom Functions using a Shared Runtime
- Excel Custom Functions using a JavaScript-only Runtime

**Comment by igor-ribeiiro** [comment]
Added the shared option before the manifest option and reordered the options for most used to least used.
The change https://github.com/OfficeDev/generator-office/pull/714/files caused a bug which the message wouldn't have the host in it:
image

So I also fixed that bug.

File: src/app/index.ts [hunk 2]
```diff
@@ -196,6 +196,9 @@ module.exports = class extends yo {
         }
       ];
       const answerForScriptType = await this.prompt(askForScriptType);
+      if (!answerForScriptType.scriptType && !this.options.ts) {
```
Comment (millerds) [inline src/app/index.ts:199]: We wanted to leave these off because the default should be set by the question settings.

File: src/app/index.ts [hunk 3]
```diff
@@ -425,10 +426,12 @@ module.exports = class extends yo {
     this.log(`NOTE: ${chalk.bgGreen('Arguments')} must be specified in the order below, and ${chalk.bgMagenta('Options')} must follow ${chalk.bgGreen('Arguments')}.\n`);
     this.log(`  ${chalk.bgGreen('projectType')}:Specifies the type of project to create. Valid project types include:`);
     this.log(`    ${chalk.yellow('angular:')}  Creates an Office add-in using Angular framework.`);
-    this.log(`    ${chalk.yellow('excel-functions:')} Creates an Office add-in for Excel custom functions.  Must specify 'Excel' as host parameter.`);
+    this.log(`    ${chalk.yellow('excel-functions-shared:')} Creates an Office add-in for Excel custom functions using a Shared Runtime.  Must specify 'Excel' as host parameter.`);
```
Comment (millerds) [inline src/app/index.ts:429]: Don't we automatically set the host parameter?  We wouldn't need this part of the message.

File: src/app/index.ts [hunk 4]
```diff
@@ -425,10 +426,12 @@ module.exports = class extends yo {
...
+    this.log(`    ${chalk.yellow('excel-functions:')} Creates an Office add-in for Excel custom functions using a JavaScript-only Runtime.  Must specify 'Excel' as host parameter.`);
     this.log(`    ${chalk.yellow('jquery:')} Creates an Office add-in using Jquery framework.`);
     this.log(`    ${chalk.yellow('manifest:')} Creates an only the manifest file for an Office add-in.`);
     this.log(`    ${chalk.yellow('react:')} Creates an Office add-in using React framework.\n`);
+    this.log(`    ${chalk.yellow('teams-manifest:')} Creates Outlook Add-in with Teams Manifest (Developer preview).\n`);
```
Comment (millerds) [inline src/app/index.ts:434]: Why is this listed as an add?  Shouldn't it already be there?

**Review by millerds — CHANGES_REQUESTED** [review]

File: src/app/index.ts (hunk 3 above)
Comment (igor-ribeiiro) [reply src/app/index.ts:429]: I guess it's because the messages are a bit different between themselves. We would need to add a new parameter to all hosts (like `displayMessage`) if we would like to keep current messages.

File: src/app/index.ts (hunk 2 above)
Comment (millerds) [reply src/app/index.ts:199]: I'm not seeing that in the documentation and if it really didn't have the default answer then wouldn't there be an error when trying to create the project since there wouldn't be an answer (i.e. something like a null exception)?

**Review by millerds — CHANGES_REQUESTED** [review]

**Comment by igor-ribeiiro** [comment]
@millerds "I'm not seeing that in the documentation and if it really didn't have the default answer then wouldn't there be an error when trying to create the project since there wouldn't be an answer (i.e. something like a null exception)?"

It is null that gets returned (or something close to it). We don't get an error because we check if it exists:
```
        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
```

The `default` is only the value that will get chosen when you press "enter", without choosing a particular option.

**Comment by millerds** [comment]
So if that line is taking care of the question not asked case, then there is no need for that if block.  It looks like the bug you are saying you are fixing is not with the language chosen, but rather the host (that is what is missing from the message).  The host also has a fall back in the same place as the language, but that isn't using the ?: case but rather the || syntax and probably behaves differently.

The point is that we shouldn't need two places to fall back onto and I prefer the _configureProject method with a simple value precedent check rather than having to maintain multiple conditionals with the same tests (the one for weather or not the question should be asked, and the same one for weather or not the question was asked.

**Comment by igor-ribeiiro** [comment]
> Member

Sure, we can handle all these cases handling at the same place. I made some changes so that all cases are covered in a single line and hopefully it is not too confusing.

File: src/app/index.ts [hunk 5]
```diff
@@ -295,12 +290,12 @@ module.exports = class extends yo {
...
         folder: this.options.output || answerForName.name || this.options.name,
         name: this.options.name || answerForName.name,
-        host: this.options.host || answerForHost.host,
         projectType: _.toLower(this.options.projectType) || _.toLower(answerForProjectType.projectType),
         isManifestOnly: isManifestProject,
         isExcelFunctionsProject: isExcelFunctionsProject,
-        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
       };
+      this.project.host = answerForHost.host ? answerForHost.host : this.options.host ? this.options.host : jsonData.getHostTemplateNames(this.project.projectType)[0];
+      this.project.scriptType = answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : this.options.js ? this.options.js : jsonData.getSupportedScriptTypes(this.project.projectType)[0];
```
Comment (millerds) [inline src/app/index.ts:298]: I think the "this.options.js ? this.options.js" is intended to be "this.options.js ? javascript"

Could also use some formatting to encourage better readability rather than all on one line.

File: src/app/index.ts [hunk 6]
```diff
@@ -29,6 +29,7 @@ let language;
 const manifest = 'manifest';
 const sso = 'single-sign-on';
 const typescript = `TypeScript`;
+let jsonData;
```
Comment (millerds) [inline src/app/index.ts:32]: Why pull this out of the try block?

File: src/app/index.ts [hunk 7]
```diff
@@ -295,12 +290,12 @@ module.exports = class extends yo {
       this.project = {
         folder: this.options.output || answerForName.name || this.options.name,
         name: this.options.name || answerForName.name,
-        host: this.options.host || answerForHost.host,
         projectType: _.toLower(this.options.projectType) || _.toLower(answerForProjectType.projectType),
         isManifestOnly: isManifestProject,
         isExcelFunctionsProject: isExcelFunctionsProject,
-        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
       };
+      this.project.host = answerForHost.host ? answerForHost.host : this.options.host ? this.options.host : jsonData.getHostTemplateNames(this.project.projectType)[0];
```
Comment (millerds) [inline src/app/index.ts:297]: Why is this moved out the block it was in?

**Review by millerds — CHANGES_REQUESTED** [review]

File: src/app/index.ts (hunk 6 above)
Comment (igor-ribeiiro) [reply src/app/index.ts:32]: It is being used on another function. So I wanted to let this other function `configureProject` also have access to `jsonData`

File: src/app/index.ts [hunk 8]
```diff
@@ -289,12 +290,17 @@ module.exports = class extends yo {
       this.project = {
         folder: this.options.output || answerForName.name || this.options.name,
         name: this.options.name || answerForName.name,
-        host: this.options.host || answerForHost.host,
         projectType: _.toLower(this.options.projectType) || _.toLower(answerForProjectType.projectType),
         isManifestOnly: isManifestProject,
         isExcelFunctionsProject: isExcelFunctionsProject,
-        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
       };
+      this.project.host = answerForHost.host ? answerForHost.host
```
Comment (millerds) [inline src/app/index.ts:297]: Why not leave this in the block using "host: answerforHost.host . . ." with everything else (you don't need to use "this.project.")?

File: src/app/index.ts [hunk 9]
```diff
@@ -289,12 +290,17 @@ module.exports = class extends yo {
...
         name: this.options.name || answerForName.name,
-        host: this.options.host || answerForHost.host,
         projectType: _.toLower(this.options.projectType) || _.toLower(answerForProjectType.projectType),
         isManifestOnly: isManifestProject,
         isExcelFunctionsProject: isExcelFunctionsProject,
-        scriptType: answerForScriptType.scriptType ? answerForScriptType.scriptType : this.options.ts ? typescript : javascript
       };
+      this.project.host = answerForHost.host ? answerForHost.host 
+        : this.options.host ? this.options.host 
+        : jsonData.getHostTemplateNames(this.project.projectType)[0];
```
Comment (millerds) [inline src/app/index.ts:299]: If this is coming from outside of this method then we should check that it's not empty.

**Review by millerds — CHANGES_REQUESTED** [review]

File: src/app/index.ts (hunk 8 above)
Comment (millerds) [reply src/app/index.ts:297]: They set projectType inside the block as well.  It's a big deal but I would add 'const projType = _.toLower(this.options.projectType) || _.toLower(answerForProjectType.projectType)' before the block then create everything inside the block using the projType variable.

File: src/app/index.ts [hunk 10]
```diff
@@ -287,20 +287,35 @@ module.exports = class extends yo {
...
   _configureProject(answerForProjectType, answerForScriptType, answerForHost, answerForName, isManifestProject, isExcelFunctionsProject): void {
     try {
+      const projType = _.toLower(this.options.projectType) || _.toLower(answerForProjectType.projectType)
+
       this.project = {
         folder: this.options.output || answerForName.name || this.options.name,
+        host: answerForHost.host
+          ? answerForHost.host
+          : this.options.host
+          ? this.options.host
```
Comment (millerds) [inline src/app/index.ts:297]: Add a new indent level here.

File: src/app/index.ts [hunk 11]
```diff
@@ -287,20 +287,35 @@ module.exports = class extends yo {
...
+          : jsonData.getHostTemplateNames(projType)[0],
         name: this.options.name || answerForName.name,
-        projectType: _.toLower(this.options.projectType) || _.toLower(answerForProjectType.projectType),
+        projectType: projType,
+        scriptType: answerForScriptType.scriptType
+          ? answerForScriptType.scriptType
+          : this.options.ts
+          ? typescript
```
Comment (millerds) [inline src/app/index.ts:304]: Add an indent level here.

File: src/app/index.ts [hunk 12]
```diff
@@ -287,20 +287,35 @@ module.exports = class extends yo {
...
+          : this.options.js
+          ? javascript
```
Comment (millerds) [inline src/app/index.ts:306]: Add an indent level here.

**Review by millerds — CHANGES_REQUESTED** [review]

File: src/app/index.ts (hunk 12 above)
Comment (igor-ribeiiro) [reply src/app/index.ts:306]: I'm using the prettier here. So those two spaces are the recommended amount.

File: src/app/index.ts (hunk 11 above)
Comment (igor-ribeiiro) [reply src/app/index.ts:304]: Same as other comment, I'm using `format selection` from vscode

File: src/app/index.ts (hunk 10 above)
Comment (igor-ribeiiro) [reply src/app/index.ts:297]: Using vscode format selection

File: src/app/index.ts [hunk 13]
```diff
@@ -287,20 +287,35 @@ module.exports = class extends yo {
...
       };
-      this.project.host = answerForHost.host ? answerForHost.host 
-        : this.options.host ? this.options.host 
-        : jsonData.getHostTemplateNames(this.project.projectType)[0];
-      this.project.scriptType = answerForScriptType.scriptType ? answerForScriptType.scriptType
-        : this.options.ts ? typescript
-        : this.options.js ? javascript
-        : jsonData.getSupportedScriptTypes(this.project.projectType)[0];
+
+      if (!this.project.host) {
```
[the remark this answers is not in the document]
Comment (igor-ribeiiro) [reply src/app/index.ts:312]: It should never be null unless there is an error with our code. I will just check for `jsonData` null

**Review by millerds — CHANGES_REQUESTED** [review]

File: src/app/index.ts [hunk 14]
```diff
@@ -295,7 +295,7 @@ module.exports = class extends yo {
           ? answerForHost.host
           : this.options.host
           ? this.options.host
-          : jsonData.getHostTemplateNames(projType)[0],
+          : jsonData?.getHostTemplateNames(projType)[0],
```
[the remark this answers is not in the document]
Comment (igor-ribeiiro) [reply src/app/index.ts:298]: It would get host = `null` which would error out later on the code in the `npm run convertToSingleHost` line. 
JSON data is currently never null. If it were to be null, then there is an with our code itself not related to a client's path of choices.

In that case, do you think throwing an error if it is null would help? Ideally, this error never appears, but it might happen if we make mistakes in our code in the future.

[the remark this answers is not in the document]
Comment (millerds) [reply src/app/index.ts:298]: It's probably OK as is.  You are right in that it's a safety for us rather than customers.  I just wanted to make sure there wouldn't be some other cryptic error or silent failure if null was encountered.

**Review by millerds — DISMISSED** [review]

**Comment by igor-ribeiiro** [comment]
@millerds could you review it again? Needed to comment some SSO tests that were added back during the merge

**Review by millerds — APPROVED** [review]
